### PR TITLE
Add loop and other control flow expressions 

### PR DESCRIPTION
**Pull request by xFrednet**
This PR adds loop expressions and control flow expressions to Marker's API. With this `cargo dogfood` only complains about the path resolution of two generic parameters (which I'm planning to fix soon) and closures. The rest seems to be working :tada:.

For and While loops again have some interesting desugar, I tried to document them in the code. It seems to be working well, at least according to the test output.

Most of the changes should be test output and documentation as usual. Sorry, that it involves so many changes at once. (I'm totally not using Marker to procrastinate) :sweat_smile: 

---

r? @Niki4tap If you have the time, I'd appreciate your feedback :)

cc: #52

Not much more to say. For everyone reading this, have a wonderful day :)

### CODE REVIEW

File: marker_api/src/ast/expr/control_flow_expr.rs [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+impl<'ast> ReturnExpr<'ast> {
+    pub fn new(data: CommonExprData<'ast>, expr: Option<ExprKind<'ast>>) -> Self {
+        Self {
+            data,
+            expr: expr.into(),
+        }
+    }
+}
+
+/// A break expression with an optional label and value.
```
Comment (Niki4tap) [inline marker_api/src/ast/expr/control_flow_expr.rs:283]:
```suggestion
/// A break expression with an optional label and a value.
```

File: marker_api/src/ast/expr/block_expr.rs [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+/// fn foo() {}
+///
+/// //      vvvvvvv An unsafe block
+/// let _ = unsafe {
+///     1 + 2
+/// //  ^^^^^ An expression which value is returned from the block, indicated
+/// //        by the missing semicolon at the end.
+/// };
+///
+/// //      vvvvvv A block expression with an optional label.
```
Comment (Niki4tap) [inline marker_api/src/ast/expr/block_expr.rs:23]: I think the doc from below `ForExpr` might be a bit better for this:
> `An optional label to be targeted by break and continue expressions`

**Review by Niki4tap — COMMENTED** [review]
Sorry for the delay ^^

Commenting mostly on the docs, as I haven't checked the implementation very thoroughly, but the tests seem to be working, so I *think* it should be fine (famous last words) :sweat_smile:

**Comment by xFrednet** [comment]
> Sorry for the delay ^^

No problem at all, thank you for providing feedback!! I have a busy week ahead. So, it might take some time until I address the comments :)

> Commenting mostly on the docs, as I haven't checked the implementation very thoroughly, but the tests seem to be working, so I _think_ it should be fine (famous last words) :sweat_smile: 

That's totally fine. The docs and API are the important parts. I find the rustc backend a bit chaotic. While working on it, I also focus on the tests. That's what they're for. So feel free to continue handling it like that :).

File: marker_api/src/ast/expr/control_flow_expr.rs [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+#[derive(Debug)]
+pub struct ForExpr<'ast> {
+    data: CommonExprData<'ast>,
+    label: FfiOption<Ident<'ast>>,
+    pat: PatKind<'ast>,
+    // FIXME: I can't find a better name for this expression. The `iter()`
+    // function name wouldn't work, as that is reserved by conventions in Rust.
+    // `values` or something similar also doesn't quite fit IMO and the reference
+    // also doesn't provide a good name.
+    iter_expr: ExprKind<'ast>,
```
[the remark this answers is not in the document]
Comment (xFrednet) [reply marker_api/src/ast/expr/control_flow_expr.rs:525]: Interesting suggestion. When I read `iterable` my first thought is that it's a property of something and not necessary the expression. But in this context that might work. I'll think a bit more about it :)

File: marker_api/src/ast/expr/block_expr.rs [hunk 4]
```diff
@@ -1,16 +1,42 @@
 use crate::{
-    ast::stmt::StmtKind,
+    ast::{stmt::StmtKind, Ident},
     ffi::{FfiOption, FfiSlice},
 };
 
 use super::{CommonExprData, ExprKind};
 
+/// A block expression, is one of the most fundamental expressions in Rust. It
```
[the remark this answers is not in the document]
Comment (xFrednet) [reply marker_api/src/ast/expr/block_expr.rs:8]: I think the comma was wrong :+1:

File: marker_api/src/ast/expr/control_flow_expr.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (xFrednet) [reply marker_api/src/ast/expr/control_flow_expr.rs:525]: I like it more than `iter_expr` and updated it in the code now. Thank you for the suggestion :)

**Comment by xFrednet** [comment]
I hope the last commit covers all suggestions. Thank you for the review :)

**Comment by Niki4tap** [comment]
Thanks! Looks great now!

bors r+

**Pull request merged**